src/operator/tensor/histogram.cc

```diff
@@ -0,0 +1,162 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "./histogram-inl.h"
+
+namespace mxnet {
+namespace op {
+
+struct ComputeBinKernel {
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(int i, const DType* in_data, const DType* bin_bounds,
+                                  int* bin_indices, int bin_cnt, double min, double max) {
+    DType data = in_data[i];
+    int target = -1;
+    if (data >= min && data <= max) {
+      target = (data - min) * bin_cnt / (max - min);
+      target = mshadow_op::minimum::Map(bin_cnt - 1, target);
+      target -= (data < bin_bounds[target]) ? 1 : 0;
+      target += ((data >= bin_bounds[target + 1]) && (target != bin_cnt - 1)) ? 1 : 0;
+    }
+    bin_indices[i] = target;
+  }
+
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(int i, const DType* in_data, int* bin_indices,
+                                   const DType* bin_bounds, int num_bins) {
+    DType data = in_data[i];
+    int target = -1;
+    if (data >= bin_bounds[0] && data <= bin_bounds[num_bins]) {
+      target = 0;
+      while ((data - bin_bounds[target]) >= 0) {
+        target += 1;
+      }
+      target = mshadow_op::minimum::Map(target - 1, num_bins - 1);
+    }
+    bin_indices[i] = target;
+  }
+};
+
+template<typename CType>
+void ComputeHistogram(const int* bin_indices, CType* out_data, size_t input_size) {
+  for (size_t i = 0; i < input_size; ++i) {
+    int target = bin_indices[i];
+    if (target >= 0) {
+      out_data[target] += 1;
+    }
+  }
+}
+
+template<typename cpu>
```

Comment:
Should this be a template specialization? Same for other places.
```cpp
template<>
void HistogramForwardImpl<cpu>()
```

Reply:
Done